lib/media-controls-impl.js

```diff
@@ -0,0 +1,524 @@
+  this.shadowRoot.addEventListener('touchstart', this);
+  this.shadowRoot.addEventListener('touchmove', this);
+  this.shadowRoot.addEventListener('touchend', this);
+  this.shadowRoot.addEventListener('mousedown', this);
+
+  this.mediaPlayer.addEventListener('loadedmetadata', this);
+  this.mediaPlayer.addEventListener('play', this);
+  this.mediaPlayer.addEventListener('pause', this);
+  this.mediaPlayer.addEventListener('timeupdate', this)
+  this.mediaPlayer.addEventListener('seeked', this);
+  this.mediaPlayer.addEventListener('ended', this);
+};
+
+MediaControlsImpl.prototype.removeEventListeners = function() {
+  this.shadowRoot.removeEventListener('touchstart', this);
+  this.shadowRoot.removeEventListener('touchmove', this);
+  this.shadowRoot.removeEventListener('touchend', this);
+  this.shadowRoot.removeEventListener('mousedown', this);
+
+  this.mediaPlayer.removeEventListener('loadedmetadata', this);
+  this.mediaPlayer.removeEventListener('play', this);
+  this.mediaPlayer.removeEventListener('pause', this);
+  this.mediaPlayer.removeEventListener('timeupdate', this)
+  this.mediaPlayer.removeEventListener('seeked', this);
+  this.mediaPlayer.removeEventListener('ended', this);
+
+  if (this.mouseEventHandlerRegistered) {
+    this.shadowRoot.removeEventListener('mousemove', this, true);
+    this.shadowRoot.removeEventListener('mouseup', this, true);
+  }
+
+  this.mouseEventHandlerRegistered = false;
+};
+
+MediaControlsImpl.prototype.handleEvent = function(e) {
+
+  // If we get a touchstart, don't process mouse events
+  if (e.type === 'touchstart') {
+    this.shadowRoot.removeEventListener('mousedown', this);
+    this.shadowRoot.removeEventListener('mousemove', this);
+    this.shadowRoot.removeEventListener('mouseup', this);
+  }
+
+  switch(e.type) {
+
+    case 'mousedown':
+      //
+      // The component is listening to window 'mousemove' events so
+      // that the slider movement will function even when the mouse
+      // moves off the play head. However, if the component is always
+      // listening to the window events, it would receive very many
+      // spurious 'mousemove' events. To prevent this, the component
+      // only listens to 'mousemove' events after receiving a 'mousedown'
+      // event.
+      //
+      if (!this.mouseEventHandlerRegistered) {
+        window.addEventListener('mousemove', this, true);
+        window.addEventListener('mouseup', this, true);
+        this.mouseEventHandlerRegistered = true;
+      }
+
+      // fall through to touchstart...
+
+    case 'touchstart':
+      switch(e.target) {
+        case this.els.play:
+          //
+          // Let the 'play' and 'pause' handlers take care of changing
+          // the icon and setting the l10n-id (for the screen reader).
+          //
+          if (this.mediaPlayer.paused) {
+            this.mediaPlayer.play();
+          }
+          else {
+            this.mediaPlayer.pause();
+          }
```

Comment:
Is this what we do currently in the music app? Does the play/pause button really respond on touchstart instead of on click or touchend?  It seems wrong that one of the three buttons behaves differently than the other two.  Is there a reason for that?